File: Java/problem4-1.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        int m = nums1.length;
        int n = nums2.length;
        if (m == 0) {
            return n % 2 == 0 ? (nums2[n/2] + nums2[n/2 - 1]) / 2.0 : nums2[n/2];
        }
        if (n == 0) {
            return m % 2 == 0 ? (nums1[m/2] + nums1[m/2 - 1]) / 2.0 : nums1[m/2];
        }
        int idx1 = 0, idx2 = 0;
        int count = 0;
        int middle = 0;
        while (count * 2 < m + n) {
            if (idx2 >= n || (idx1 < m && nums1[idx1] < nums2[idx2])) {
                middle = nums1[idx1++];       
            }
            else {
                middle = nums2[idx2++];
            }
            ++count;
        }
        if ((m + n) % 2 == 1) {
            return middle * 1.0;
        }
        else {
            int temp = (idx2 >= n || (idx1 < m && nums1[idx1] < nums2[idx2])) ? nums1[idx1] : nums2[idx2];
            return (middle + temp) / 2.0;
        }
    }
}

File: python/problem4-1.py
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        m = len(nums1)
        n = len(nums2)
        if m > n:
            return self.findMedianSortedArrays(nums2, nums1)
        if m == 0:
            return (nums2[n//2 - 1] + nums2[n//2]) / 2 if n % 2 == 0 else nums2[n//2]
        iMin, iMax = 0, m
        middle = (m + n + 1) // 2
        while iMin <= iMax:
            i = iMin + (iMax - iMin) // 2
            j = middle - i
            if i > 0 and nums1[i - 1] > nums2[j]:
                iMax = i - 1
            elif i < m and nums1[i] < nums2[j - 1]:
                iMin = i + 1
            else:
                middleNum, middleNum_1 = 0, 0
                if i == 0:
                    middleNum = nums2[j - 1]
                elif j == 0:
                    middleNum = nums1[i - 1]
                else:
                    middleNum = max(nums1[i - 1], nums2[j - 1])
                if (m + n) % 2 == 1:
                    return middleNum
                if i == m:
                    middleNum_1 = nums2[j]
                elif j == n:
                    middleNum_1 = nums1[i]
                else:
                    middleNum_1 = min(nums1[i], nums2[j])
                return (middleNum + middleNum_1) / 2
        return 0.0
